test/boost/sstable_set_test.cc

```diff
@@ -0,0 +1,269 @@
+        auto key_and_token_pair = token_generation_for_current_shard(2);
+        auto sst = sstable_for_overlapping_test(env, s, 1, key_and_token_pair[0].first, key_and_token_pair[0].first, 1);
+        set->insert(sst);
+        // set' -> list
+        //      -> list2
+        //      -> set
+        BOOST_REQUIRE_EQUAL(set->all()->version().get_previous_version(), list->version().get_previous_version());
+        BOOST_REQUIRE_EQUAL(set->all()->version().get_previous_version(), list2->version().get_previous_version());
+        BOOST_REQUIRE_NE(&set->all()->version(), list->version().get_previous_version());
+        sst = sstable_for_overlapping_test(env, s, 1, key_and_token_pair[1].first, key_and_token_pair[1].first, 1);
+        set->insert(sst);
+        // set' -> list
+        //      -> list2
+        //      -> set
+        BOOST_REQUIRE_EQUAL(set->all()->version().get_previous_version(), list->version().get_previous_version());
+        BOOST_REQUIRE_EQUAL(set->all()->version().get_previous_version(), list2->version().get_previous_version());
+        BOOST_REQUIRE_NE(&set->all()->version(), list->version().get_previous_version());
+        sstable_list_opt list3 = list;
+        // set' -> list -> list3
+        //      -> list2
+        //      -> set
+        BOOST_REQUIRE_EQUAL(&list->version(), list3->version().get_previous_version());
+        list = std::nullopt;
+        // set' -> list3
+        //      -> list2
+        //      -> set
+        BOOST_REQUIRE_EQUAL(set->all()->version().get_previous_version(), list3->version().get_previous_version());
+        BOOST_REQUIRE_EQUAL(list2->version().get_previous_version(), list3->version().get_previous_version());
+        set = std::nullopt;
+        // set' -> list3
+        //      -> list2
+        BOOST_REQUIRE_EQUAL(list2->version().get_previous_version(), list3->version().get_previous_version());
+        sstable_list_opt list4 = list3;
+        sstable_list_opt list5 = list3;
+        // set' -> list3 -> list4
+        //               -> list5
+        //      -> list2
+        BOOST_REQUIRE_EQUAL(&list3->version(), list4->version().get_previous_version());
+        BOOST_REQUIRE_EQUAL(&list3->version(), list5->version().get_previous_version());
+        list3 = std::nullopt;
+        // set' -> list3' -> list4
+        //                -> list5
+        //      -> list2
+        BOOST_REQUIRE_NE(list2->version().get_previous_version(), list4->version().get_previous_version());
+        BOOST_REQUIRE_EQUAL(list4->version().get_previous_version(), list5->version().get_previous_version());
+        list4 = std::nullopt;
+        // set' -> list5
+        //      -> list2
+        BOOST_REQUIRE_EQUAL(list2->version().get_previous_version(), list5->version().get_previous_version());
+        list2 = std::nullopt;
+        // list5
+        BOOST_REQUIRE_EQUAL(nullptr, list5->version().get_previous_version());
+        return make_ready_future<>();
+    });
+}
```

Comment:
I'd like to see a random walk test where the test generates random operations (add sstable, remove sstable, create version, delete version, get all, select, etc) and applies the random operations to both the new code, and to simpler code that uses copies to create versions, comparing the results for read operations. See chunked_vector_test for an example.

Reply:
You can see it in the case `sstable_set_random_walk_test`